### ISSUE TITLE
Occasional "wait: no child processes" errors when starting VMs

### ISSUE THREAD

**Issue by sipsma** [firecracker-microvm/firecracker-containerd#285]
I've occasionally gotten this error when trying to spin up containers/VMs:
```
ctr: aws.firecracker: unexpected error from CreateVM: rpc error: code = Unknown desc = shim CreateVM returned error: rpc error: code = Unknown desc = failed to create VM: failed to start the VM: failure when invoking CNI: failed to delete pre-existing CNI network {NetworkName:fcnet IfName:veth0 Args:[] BinPath:[/opt/cni/bin] ConfDir:/etc/cni/conf.d CacheDir:/var/lib/cni/b11bb660-35f8-4336-9175-9c9456c83122 containerID:b11bb660-35f8-4336-9175-9c9456c83122 netNSPath:/var/run/netns/b11bb660-35f8-4336-9175-9c9456c83122}: failed to delete CNI network list "fcnet": wait: no child processes
```

Usually it will succeed on retry but it's extremely inconsistent. I've at times encountered it about half the time I try to spin up containers, other times seemingly unrelated changes make it extremely rare. I've managed to see it twice in integ tests, but only when modifying them locally to spin up many more VMs concurrently.

Looking into root error `wait: no child processes`, I actually don't think it has to do with CNI specifically. The error appears to be returned directly from the `os/exec` package's `Command.Run` method [here](https://github.com/containernetworking/cni/blob/63202159f106dd4f4f0ae3e155649b5775946268/pkg/invoke/raw_exec.go#L39). The error message suggests that for some reason the child process [has already been waited on between the internal call to `Start` and internal call to `Wait`](https://github.com/golang/go/blob/33fac9bad86d7f3bba8926d92c33e5a0cf28061f/src/os/exec/exec.go#L337-L342). It thus seems more likely that there is some general issue with creating child processes which just became more apparent when adding CNI support because CNI plugins result in many more child processes being spawned in rapid succession than we had done previously.

In terms of what could cause issues with spawning child processes, I first tried [disabling setting the shim as a subreaper in this now closed PR](https://github.com/firecracker-microvm/firecracker-containerd/pull/284), which at first seemed to work but turned out to just make it more rare in practice (unknown why, it's possibly just a red herring).

I don't have any other strong suspicions at this moment but one thing worth looking into is possible interactions between [the code in `github.com/containerd/containerd/sys` here](https://github.com/containerd/containerd/blob/master/sys/subprocess_unsafe_linux.go) and the error we're getting. We do import that `github.com/containerd/containerd/sys` package in our shim indirectly due to our [import here](https://github.com/firecracker-microvm/firecracker-containerd/blob/master/runtime/main.go#L20). Need to understand what is happening with the relinking of the golang internal fork functions better though, it could be completely unrelated.

This will also need to be reproducible more consistently. It's possible just temporarily modifying the shim to do something silly like spawn a bunch of `true` subprocesses would be good enough for testing any possible fixes.

**Comment by sipsma** [comment]
1. I modified the go sdk to run 128 `/bin/true` child processes right before it does the CNI calls. This results in fairly consistently getting the same `wait: no child process` error but for the `true` processes, so this seems to confirm my suspicion that this is unrelated to CNI and is instead a problem in general with spawning child processes.
2. Furthering the mystery though I then tried modifying firecracker-containerd to spawn a bunch of `true` child procs at the beginning of [the NewService function](https://github.com/firecracker-microvm/firecracker-containerd/blob/65d35319a4fcd95941af285279f4aac4240527fa/runtime/service.go#L140) instead of spawning them in the go sdk. I never got any `wait no child process` errors from the child procs in `NewService` over many runs (though it's of course impossible to say whether it can never occur in `NewService` or it's just more rare there).

The above **may** suggest that whatever is causing this issue only occurring later in process execution (or is somehow specific to code in the Go sdk, though I don't know how that could ever happen).

**Comment by sipsma** [comment]
Ah okay, I believe I found the issue. I missed this last time I looked at the code, but there is both a `NoSubreaper` **AND** a `NoReaper` option in the containerd shim: https://github.com/containerd/containerd/blob/08061c7c3cd095e4d35fc7999c21dc7bad8a9623/runtime/v2/shim/shim.go#L77-L80

I only noticed the `NoSubreaper` option last attempt at fixing this, but the most important thing for us to disable is the `NoReaper` option, which is what actually sets up the `SIGCHLD` handler loop (which then races with golang's `os/exec` wait on child processes)...

The reason that spawning `/bin/true` in the `NewService` function didn't result in any error is because `NewService` is [run before](https://github.com/containerd/containerd/blob/08061c7c3cd095e4d35fc7999c21dc7bad8a9623/runtime/v2/shim/shim.go#L182) the reaper loop is setup by containerd shim code.

Setting both `NoSubreaper` and `NoReaper` has prevented me from being able to reproduce the error even when spawning tons of `/bin/true` processes, so that in combination with the logical explanation for what's happening gives me reasonable confidence this actually fixes the issue.
